### PR TITLE
Remove race conditions in ProducerPopulatedTagSource

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Previously ProducerPopulatedTagSource would compute results on the background and then mutate its internal state with those values it had computed.

Then on the foreground it might also try to mutate those values.  This was very racey, and there were comments indicating as such in the code.  This was especially egregious for the code trying to track what span of text had been edited over time.  It was not only racey in how it accumulated and used this value, it was also racing in how it reset the value once it had been used.  This could mean we might think there was no text change when there actually was, or vice-versa.

I've now switched the ProducerPopulatedTagSource  to strictly follow the pattern of

1. On change, switch to foreground. (we were already doing this).  Snapshot state and enqueue bg request.
2. On background compute new results.  Make no mutations to existing state.  Enqueue fg to update state.
3. On foreground update state.

This removes the need for any sort of locking in this type, and makes the conceptual model much simpler.  We use the FG thread effectively as the serialization mechanism to make sure all reads/writes to this state are consistent and coherent.

Note that even though we switch to the foreground at the end, all we're doing is updating our internal state, and then letting the BatchChangeNotifier know about the changes at that point.  The batch change notifier will still throttle and coalesce UI notifications to the editor about which spans have actually changed.  So it's not like this change will cause us to do undue stuff like calling into the editor too often with too many changes.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dpoeschl and @heejaechang

**Comment by CyrusNajmabadi** [comment]
And also @jasonmalinowski

File: src/EditorFeatures/Core/Implementation/ReferenceHighlighting/ReferenceHighlightingTagSource.cs [hunk 1]
```diff
@@ -123,7 +124,12 @@ private void ClearTags(List<DocumentSnapshotSpan> spansToTag, CancellationToken
             }
 
             spansToTag = spansToTag ?? GetSpansAndDocumentsToTag();
-            this.WorkQueue.EnqueueBackgroundTask(c => this.ClearTagsAsync(spansToTag, c), "NoTags", cancellationToken);
+
+            // Save to access CachedTagTrees here because we're on the foreground thread.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/ReferenceHighlighting/ReferenceHighlightingTagSource.cs]: "safe to"

**Comment by heejaechang** [comment]
@CyrusNajmabadi I would prefer either using lock or having another queue for serialization rather than using FG for such purpose.

**Comment by CyrusNajmabadi** [comment]
I’m curious what’s wrong with using the FG thread.  We already report back to that thread for multiple purposes (i.e. the batch change notifier as well as the event source notifications).

Since all work starts from the FG thread and ends back up on the FG thread, it seems like a nice place to do the state updating.

             -- Cyrus

**Comment by heejaechang** [comment]
Since FG is precious special thread, and FG should be used for something that require FG (such as UI update, STA COM API). If something that can be done without FG, it should be done without FG is my opinion.


-          heejae

**Comment by CyrusNajmabadi** [comment]
I’d be ok with that.  Except we do already come back to the UI thread to update the editor.  As such, it seems like a totally good time to actually update state (since this is a total of 3 pointer writes).

Note: my changes in the large tagging thread actually make it so we end up going back to the UI thread far less.  There are at least two times we used to go back to the UI thread that I’ve removed.  So that’s good ☺

            -- Cyrus

**Pull request merged**